File: src/Solutions/Euler055.hs
module Solutions.Euler055 where
import Utils.Numeric(reverseInt, isPalindrome)

lynchrel iter n =
    iter == 0 || let n' = reverseInt n + n
                 in not (isPalindrome n') && lynchrel (iter - 1) n'

euler055 :: IO String
euler055 = return $ show $ length $ filter (lynchrel 50) [1..9999]

File: rust/src/solutions/euler055.rs
use crate::utils::numeric::digits::Digits;
use num::bigint::BigInt;

fn palindrome(n: &BigInt) -> BigInt {
    n.digits()
        .into_iter()
        .rev()
        .fold(BigInt::from(0), |a, b| 10 * a + BigInt::from(b))
}

fn lynchrel(mut n: BigInt) -> bool {
    for _ in 0..50 {
        n = &n + palindrome(&n);
        if n.is_palindrome() {
            return false;
        }
    }
    true
}

pub fn euler055() -> String {
    (1..10000)
        .filter(|&n| lynchrel(BigInt::from(n)))
        .count()
        .to_string()
}
